sound/soc/sof/sof-audio.c

```diff
@@ -128,20 +135,27 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 		break;
 	}
 
+	/* continue to disable core even if IPC fails */
 	ret = sof_ipc_tx_message(sdev->ipc, ipc_free.hdr.cmd, &ipc_free, sizeof(ipc_free),
 				 &reply, sizeof(reply));
-	if (ret < 0) {
+	if (ret < 0)
 		dev_err(sdev->dev, "error: failed to free widget %s\n", swidget->widget->name);
-		swidget->use_count++;
-		return ret;
-	}
+
+	/*
+	 * disable widget core. continue to route setup status and complete flag
+	 * even if this fails
+	 */
+	if (snd_sof_dsp_core_put(sdev, core) < 0)
+		dev_err(sdev->dev, "error: failed to disable target core: %d for widget %s\n",
+			core, swidget->widget->name);
```

Comment:
do you also want to return 0 when such an error is reached? A comment would be good.

Reply:
or dev_warning ?